uplink/src/base/mod.rs

```diff
@@ -170,33 +198,73 @@ where
         self.last_sequence = current_sequence;
         self.last_timestamp = current_timestamp;
 
-        // Send buffer to serializer
-        if self.buffer.buffer.len() >= self.max_buffer_size {
-            let name = self.name.clone();
-            let topic = self.topic.clone();
-            let buffer = mem::replace(&mut self.buffer, Buffer::new(name, topic));
-            Ok(Some(buffer))
+        // if max_buffer_size is breached, flush
+        let buf = if self.buffer.buffer.len() >= self.max_buffer_size {
+            Some(self.take_buffer())
         } else {
-            Ok(None)
+            None
+        };
+
+        Ok(buf)
+    }
+
+    // Returns buffer content, replacing with empty buffer in-place
+    fn take_buffer(&mut self) -> Buffer<T> {
+        let name = self.name.clone();
+        let topic = self.topic.clone();
+        info!("Flushing stream name: {}, topic: {}", name, topic);
+
+        mem::replace(&mut self.buffer, Buffer::new(name, topic))
+    }
+
+    /// Triggers flush and async channel send if not empty
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        if !self.is_empty() {
+            let buf = self.take_buffer();
+            self.tx.send_async(Box::new(buf)).await?;
         }
+
+        Ok(())
+    }
+
+    /// Returns number of elements in Stream buffer
+    pub fn len(&self) -> usize {
+        self.buffer.buffer.len()
     }
 
-    /// Fill buffer with data and trigger async channel send on max_buf_size
-    pub async fn fill(&mut self, data: T) -> Result<(), Error> {
+    /// Check if Stream buffer is empty
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn stream_status(&self) -> StreamStatus<'_> {
+        if self.len() == 1 {
+            return StreamStatus::Init(&self.name, self.flush_period);
+        }
+
+        StreamStatus::Partial(self.len())
+    }
+
+    /// Fill buffer with data and trigger async channel send on breaching max_buf_size.
+    /// Returns [`StreamStatus`].
+    pub async fn fill(&mut self, data: T) -> Result<StreamStatus<'_>, Error> {
         if let Some(buf) = self.add(data)? {
             self.tx.send_async(Box::new(buf)).await?;
+            return Ok(StreamStatus::Flushed(&self.name));
         }
 
-        Ok(())
+        Ok(self.stream_status())
```

Comment:
Returning `StreamStatus` enum explicitly on line 25 and from a method here is confusing. Either repeat the code or return `StreamStatus` from `add` instead of `Option<Buffer`